src/std/db/postgresql-driver.ss

```diff
@@ -169,90 +171,118 @@
   error: "error closing connection")
 
 ;;; driver implementation
-(def (postgresql-connect! host port user passwd db)
+(def (postgresql-connect! host port user passwd db ssl ssl-context timeout)
   (def sock
     (tcp-connect (cons host port)))
 
-  (def reader
-    (open-buffered-reader
-     (StreamSocket-reader sock)))
-
-  (def writer
-    (open-buffered-writer
-     (StreamSocket-writer sock)))
+  (try
 
-  (def (start-driver!)
-    (DEBUG "STARTING DRIVER")
-    (let lp ()
+   (start-logger!)
+   (DEBUG "STARTUP")
+
+   (when ssl
+     ;; Send magic SSLRequest message with non-version 80877103, instead of StartupMessage
+     ;; https://www.postgresql.org/docs/current/protocol-flow.html
+     (StreamSocket-send sock #u8(0 0 0 8 4 210 22 47) 0 8 0)
+     (let* ((buf (make-u8vector 1 0))
+            (count (StreamSocket-recv sock buf 0 1 MSG_WAITALL))
+            (response (u8vector-ref buf 0)))
+       (cond
+        ((not (= count 1))
+         (error "Postgres connection failed immediately"))
+        ((= response 78) ;; (char->integer #\N)
+         ;; Exceptionally (see URL above), this N isn't followed by a notice message
+         (when (eq? ssl #t)
+           (error "Postgres Server does not support SSL encryption.")))
+        ((= response 83) ;; (char->integer #\S)
+         (set! sock (ssl-client-upgrade sock (make-timeout timeout #f)
+                                        context: ssl-context
+                                        host: host)))
+        (else
+         ;; read as much as we can from the server response to give the best error context
+         (let* ((buf1 (make-u8vector 2048 0))
+                (len0 (StreamSocket-recv sock buf1 1 2048 MSG_DONTWAIT))
```

Comment:
`(let* ((len ...) (len ...)) ...)`

No need to pollute the namespacr and brain context if we only use the second.